InnerEye/ML/utils/model_util.py

```diff
@@ -94,26 +94,35 @@ def mean_teacher_model(self) -> Optional[DeviceAwareModule]:
 
     @classmethod
     def _load_checkpoint(cls, model: DeviceAwareModule, checkpoint_path: Path,
-                         key_in_state_dict: str, use_gpu: bool) -> int:
+                         key_in_state_dict: str, reader: Callable) -> int:
         """
         Loads a checkpoint of a model, may be the model or the mean teacher model. Assumes the model
         has already been created, and the checkpoint exists. This does not set checkpoint epoch.
         This method should not be called externally. Use instead try_load_checkpoint_for_model
         or try_load_checkpoint_for_mean_teacher_model
         :param model: model to load weights
+        :param checkpoint_path: Path to checkpoint
         :param key_in_state_dict: the key for the model weights in the checkpoint state dict
+        :param reader: Function which takes the path and returns a dict with model and optimizer states
         :return checkpoint epoch form the state dict
         """
         logging.info(f"Loading checkpoint {checkpoint_path}")
-        # For model debugging, allow loading a GPU trained model onto the CPU. This will clearly only work
-        # if the model is small.
-        map_location = None if use_gpu else 'cpu'
-        checkpoint = torch.load(str(checkpoint_path), map_location=map_location)
+        checkpoint = reader(checkpoint_path)
+
+        try:
+            state_dict = checkpoint[key_in_state_dict]
+        except KeyError:
+            logging.info(f"Key {key_in_state_dict} not found in checkpoint")
+            return False
 
         if isinstance(model, torch.nn.DataParallel):
-            model.module.load_state_dict(checkpoint[key_in_state_dict])
+            result = model.module.load_state_dict(state_dict, strict=False)
         else:
-            model.load_state_dict(checkpoint[key_in_state_dict])
+            result = model.load_state_dict(state_dict, strict=False)
+
+        logging.info(f"Could not find the following keys in model checkpoint: {result.missing_keys}")
+        logging.info(f"Found unexpected keys in model checkpoint: {result.unexpected_keys}")
```

Comment:
each of these lines should only be printed if the respective key set is non-empty. And maybe they should go to logging.warning rather than info?